File: web/js/_tests/unanswered.test.js
// web/js/_tests/unanswered.test.js
//
// A session that has owed its next move and said nothing. Mirrors
// internal/state/unanswered_test.go: the verdict is computed twice, once in Go for the
// banners and once here for the screen, and the two must agree case by case.
//
// Lives under _tests/ for the same reason header.test.js does: web/embed.go's plain
// directory pattern excludes it from the binary.
//
// Run with: node --test web/js/_tests/unanswered.test.js

import test from "node:test";
import assert from "node:assert/strict";

// sessions.js and needs.js reach for navigator.language at import time through i18n.js.
globalThis.navigator ??= { language: "en" };

const { waiting, isWaiting, WAITING_NO, WAITING_UNKNOWN, WAITING_YES, UNANSWERED_LIMIT_MS } = await import("../needs.js");
const { rowHtml } = await import("../sessions.js");
// The dictionary is picked from the machine's locale (Node has its own navigator), so
// wording is compared through t(), never against one language's text.
const { t } = await import("../i18n.js");

const NS_PER_MS = 1e6;
const minutes = (m) => m * 60 * 1000 * NS_PER_MS;

// A session the daemon says has no question, left owing its move for the given number
// of nanoseconds -- the wire shape of state.SessionView. With detail carrying the last
// message sent to it, which is what the daemon puts there.
function unanswered(forNs, extra = {}) {
  return {
    short: "un01",
    name: "n",
    needs: "",
    state: "working",
    tempo: "active",
    detail: "<agent-message from=\"06a1f607\">are you there?</agent-message>",
    silentFor: forNs,
    unansweredFor: forNs,
    ...extra,
  };
}

const readCall = { tool: "Read", since: "2026-09-12T18:00:00Z" };

test("a session left unanswered past the limit is unknown, not no", () => {
  assert.equal(waiting(unanswered(UNANSWERED_LIMIT_MS * NS_PER_MS)), WAITING_UNKNOWN);
});

test("a session unanswered under the limit is still no", () => {
  assert.equal(waiting(unanswered((UNANSWERED_LIMIT_MS - 1000) * NS_PER_MS)), WAITING_NO);
});

test("silence in a finished turn keeps the daemon's no", () => {
  assert.equal(waiting({ needs: "", state: "working", silentFor: minutes(40) }), WAITING_NO);
});

test("words still decide however long a session has owed its move", () => {
  assert.equal(waiting(unanswered(minutes(30), { needs: "answer: Allow ssh to a new host? (yes · no)" })), WAITING_YES);
  assert.equal(waiting(unanswered(minutes(30), { needs: "rate limited" })), WAITING_NO);
});

test("an unmeasured unanswered stretch is still no", () => {
  assert.equal(waiting(unanswered(0)), WAITING_NO);
});

test("a dying session long unanswered is no", () => {
  assert.equal(waiting(unanswered(minutes(30), { dying: true })), WAITING_NO);
});

test("a long unanswered session is never counted as waiting for a person", () => {
  assert.equal(isWaiting(unanswered(minutes(30))), false);
});

test("the unanswered limit is sixteen minutes, the same number Go uses", () => {
  // internal/state/unanswered_browser_test.go reads this constant out of needs.js and
  // compares it with state.UnansweredLimit; this pins the value from the other side.
  assert.equal(UNANSWERED_LIMIT_MS, 16 * 60 * 1000);
});

test("a long unanswered session says so on its row", () => {
  const html = rowHtml(unanswered(minutes(20)));
  assert.ok(html.includes("sbadge-unknown"), "it is the same not-known badge, not a new state");
  assert.ok(html.includes("sbadge-unanswered"), "and it says which kind of not-known it is");
  assert.ok(html.includes(`>${t("waiting_unanswered")}<`), "the fact the panel has, in words");
  assert.ok(!html.includes("sbadge-waiting"), "not known is not a question to a person");
});

test("a call is named only when the transcript shows one", () => {
  const named = rowHtml(unanswered(minutes(20), { inCall: readCall }));
  assert.ok(named.includes("sbadge-incall"));
  assert.ok(named.includes("Read"), "the call is the fact the row states");
  const unnamed = rowHtml(unanswered(minutes(20)));
  assert.ok(!unnamed.includes("sbadge-incall"), "naming a call nobody can see would be a precise lie");
});

test("the unanswered badge carries no guess about why", () => {
  const html = rowHtml(unanswered(minutes(20), { inCall: readCall }));
  for (const guess of ["frozen", "stuck", "hung", "dead"]) {
    assert.ok(!html.toLowerCase().includes(guess), `the row must not claim "${guess}"; slow and frozen look the same`);
  }
});

test("the unanswered badge leaves on the next snapshot once the session speaks", () => {
  const frozen = unanswered(minutes(20), { inCall: readCall });
  const back = { ...frozen, inCall: undefined, unansweredFor: undefined, silentFor: 2 * 1e9 };
  assert.ok(rowHtml(frozen).includes("sbadge-unanswered"));
  const next = rowHtml(back);
  assert.ok(!next.includes("sbadge-unanswered"), "a row must not outlive its cause");
  assert.ok(!next.includes("sbadge-unknown"), "with the session speaking, the daemon's no stands again");
});

test("a session briefly owing its move has no badge at all", () => {
  const html = rowHtml(unanswered(minutes(3)));
  assert.ok(!html.includes("sbadge-unknown"));
  assert.ok(!html.includes("sbadge-unanswered"));
});

test("a long unanswered session is named even while its flags make it stalled", () => {
  // Seen on a live frozen session on 2026-09-12: state stayed "blocked" the whole time,
  // so after ten minutes of silence the stall tracker counts it too. A bare flag says
  // less than the unanswered stretch does -- "stalled" with the text of the last message
  // sent to it as the reason -- so the row says what is known.
  const html = rowHtml(unanswered(minutes(20), { state: "blocked" }), true);
  assert.ok(html.includes("sbadge-unanswered"), "the unanswered stretch is the more specific thing that can be said");
  assert.ok(!html.includes("sbadge-stalled"), "one badge, the most specific one");
});

test("a stall in words still outranks everything but a question", () => {
  const html = rowHtml(unanswered(minutes(20), { needs: "usage limit reached" }), true);
  assert.ok(html.includes("sbadge-stalled"), "words decide: this session is stalled on a limit");
  assert.ok(!html.includes("sbadge-unanswered"));
});

test("a source that never reported needs keeps its own not-reported badge", () => {
  const html = rowHtml({ short: "uu11", name: "n", state: "working", detail: "working away" });
  assert.ok(html.includes("sbadge-unknown"));
  assert.ok(!html.includes("sbadge-unanswered"), "the unanswered wording belongs only to a session seen owing its move");
});

File: internal/state/unanswered_test.go
package state

import (
	"testing"
	"time"

	"github.com/kroticw/fleetdeck/internal/daemon"
	"github.com/kroticw/fleetdeck/internal/transcript"
)

// unansweredView is a session the daemon describes as having no question outstanding,
// while its transcript shows it has owed its next move, saying nothing, for the given
// time.
func unansweredView(short string, unanswered time.Duration) SessionView {
	v := idleView(short)
	v.UnansweredFor = unanswered
	v.SilentFor = unanswered
	return v
}

// The daemon's "no question outstanding" is words, and words come from the session. A
// session that has owed its next move and said nothing for long enough -- a call that
// has not come back, a result it has not acted on, a message it has not answered --
// has not been able to update those words, so its "no" is stale. The verdict for it is
// the one #138 gave a source that says nothing: not known. Not "yes": nobody
// established that a person is needed, only that the session could not say.
func TestSessionLeftUnansweredPastTheLimitIsUnknown(t *testing.T) {
	s := unansweredView("a", UnansweredLimit)
	if got := s.Waiting(); got != daemon.Unknown {
		t.Fatalf("a session silent while owing its move for %s must be unknown, got %s", UnansweredLimit, got)
	}
}

func TestSessionUnansweredUnderTheLimitIsStillNo(t *testing.T) {
	s := unansweredView("a", UnansweredLimit-time.Second)
	if got := s.Waiting(); got != daemon.No {
		t.Fatalf("under the limit is ordinary work, got %s", got)
	}
}

// Silence while owing nothing is a turn that has ended: the session said its last
// words and nobody has spoken to it since. The daemon's "no" describes that session
// exactly; a long silence there is the silence rule's to report, not this one's.
func TestSilentSessionThatOwesNothingIsStillNo(t *testing.T) {
	s := idleView("a")
	s.SilentFor = 3 * UnansweredLimit
	if got := s.Waiting(); got != daemon.No {
		t.Fatalf("a finished turn is not an unanswered one: got %s", got)
	}
}

// Words outrank everything else, as they already do the flags: a question the daemon
// relays is a yes, and a stall it names is a no, however long the session has owed.
func TestWordsStillDecideWhenLongUnanswered(t *testing.T) {
	asking := unansweredView("a", 2*UnansweredLimit)
	asking.Needs = daemon.Says("answer: Allow ssh to connect to a new host? (yes · no)")
	if got := asking.Waiting(); got != daemon.Yes {
		t.Fatalf("a question in words is yes, got %s", got)
	}

	limited := unansweredView("b", 2*UnansweredLimit)
	limited.Needs = daemon.Says("rate limited")
	if got := limited.Waiting(); got != daemon.No {
		t.Fatalf("a named stall is no, got %s", got)
	}
}

func TestDyingSessionLongUnansweredIsNo(t *testing.T) {
	s := unansweredView("a", 2*UnansweredLimit)
	s.Dying = true
	if got := s.Waiting(); got != daemon.No {
		t.Fatalf("a session being killed needs no one, got %s", got)
	}
}

func TestSourceWithoutNeedsStaysUnknownHoweverLongUnanswered(t *testing.T) {
	s := unansweredView("a", time.Minute)
	s.Needs = nil
	if got := s.Waiting(); got != daemon.Unknown {
		t.Fatalf("a source that never speaks needs is unknown already, got %s", got)
	}
}

// The call is a name for what is unanswered, never a condition for the verdict: a
// session frozen inside a Read has no call on disk at all, and must reach the same
// answer as one frozen inside a Bash that does.
func TestAVisibleCallIsNotRequiredForTheVerdict(t *testing.T) {
	seen := unansweredView("a", UnansweredLimit)
	seen.InCall = &transcript.Call{Tool: "Bash", Since: observedAt.Add(-UnansweredLimit)}
	unseen := unansweredView("b", UnansweredLimit)
	if seen.Waiting() != daemon.Unknown || unseen.Waiting() != daemon.Unknown {
		t.Fatalf("with or without a call on disk the answer is the same: %s / %s", seen.Waiting(), unseen.Waiting())
	}
}

// The banner side follows #138 unchanged: an unknown verdict raises no waiting banner of
// its own, and the silence rule -- measured from what the session said, so a message
// sent to it cannot reset it -- is what calls a person.
func TestLongUnansweredRaisesNoWaitingBannerButSilenceStillCalls(t *testing.T) {
	prev := Snapshot{At: observedAt, Sessions: []SessionView{idleView("a")}}
	next := Snapshot{Sessions: []SessionView{unansweredView("a", 2*time.Hour)}}
	fire, _ := Diff(prev, next, time.Hour)

	var silent bool
	for _, e := range fire {
		if e.Key == "session:a:waiting" {
			t.Fatalf("not known is not yes, so no waiting banner: %+v", fire)
		}
		if e.Key == "session:a:silent" {
			silent = true
		}
	}
	if !silent {
		t.Fatalf("a session silent past the threshold must still call someone: %+v", fire)
	}
}

// The limit is a measurement with two walls, and this pins both. Below: the longest a
// working session was seen owing its move on disk over the fourteen days to
// 2026-09-13, 651 seconds, and the worst case its parts can add up to -- a Bash at its
// 601-second ceiling inside a batch Claude Code writes only when it ends, after the
// model's own 338 seconds at p99.99 -- 939 seconds. Above: the thirty-minute silence
// default, which this exists to beat.
func TestUnansweredLimitSitsBetweenTheLongestWorkingStretchAndTheSilenceDefault(t *testing.T) {
	if UnansweredLimit <= 939*time.Second {
		t.Fatalf("the limit must clear the worst working stretch (939s), got %s", UnansweredLimit)
	}
	if UnansweredLimit >= 30*time.Minute {
		t.Fatalf("the limit must beat the thirty-minute silence default, got %s", UnansweredLimit)
	}
}
